File: internal/pkg/outcome/outcome.go
package outcome

import (
	"fmt"
	"sort"

	"github.com/marcelocantos/flac/internal/pkg/pinyin"
	"github.com/marcelocantos/flac/internal/pkg/proto/refdata"
)

type Outcome struct {
	Word     string
	Entries  *refdata.CEDict_Entries
	Good     pinyin.Alts
	Bad      pinyin.Alts
	TooShort pinyin.Alts
	Missing  int
	Easy     bool
}

func (o *Outcome) Pass() bool {
	return len(o.Bad)+len(o.TooShort)+o.Missing == 0
}

func (o *Outcome) Correction() string {
	return fmt.Sprintf("%s = %s", o.Word, o.WordAlts().ColorString())
}

func (o *Outcome) WordAlts() pinyin.Alts {
	wordAlts := make(pinyin.Alts, 0, len(o.Entries.Definitions))
	for raw := range o.Entries.Definitions {
		word, err := pinyin.NewWord(raw)
		if err != nil {
			panic(err)
		}
		wordAlts = append(wordAlts, word)
	}
	sort.Sort(wordAlts)
	return wordAlts
}

File: electron/src/outcome/Outcome.ts
import Alts from '../pinyin/Alts';
import Word from '../pinyin/Word';
import { Entries } from '../refdata/Refdata';

export default class Outcome {
  Good:     Word[] = [];
  Bad:      Word[] = [];
  TooShort: Word[] = [];
  BadTones: Word[] = [];
  Missing = 0;
  Easy = false;
  Score?: number;

  constructor(
    public Word:    string,
    public Entries: Entries,
    public Answer:  string,
  ){}

  get 及格(): boolean {
    return this.Bad.length + this.TooShort.length + this.BadTones.length + this.Missing === 0;
  }

  get 不及格(): boolean {
    return this.Bad.length + this.BadTones.length > 0;
  }

  get WordAlts(): Alts {
    return new Alts(Object.keys(this.Entries.entries).map(raw => new Word(raw)).sort());
  }
}
